packages/hydrogen-codegen/tests/codegen.test.ts

```diff
@@ -0,0 +1,210 @@
+import {describe, it, expect} from 'vitest';
+import path from 'path';
+import {preset, schema, pluckConfig, patchGqlPluck} from '../src/index.js';
+
+const getCodegenOptions = (fixture: string, output = 'out.d.ts') => ({
+  pluckConfig: pluckConfig as any,
+  generates: {
+    [output]: {
+      preset,
+      schema,
+      documents: path.join(__dirname, `fixtures/${fixture}`),
+    },
+  },
+});
+
+describe('Hydrogen Codegen', async () => {
+  // Patch dependency before importing the Codegen CLI
+  await patchGqlPluck();
+  const {executeCodegen} = await import('@graphql-codegen/cli');
+
+  it('requires .d.ts extension', async () => {
+    await expect(
+      executeCodegen(getCodegenOptions('simple-operations.ts', 'out')),
+    ).rejects.toThrowErrorMatchingInlineSnapshot(
+      '"[hydrogen-preset] target output should be a .d.ts file"',
+    );
+  });
+
+  it('prevents duplicate operations', async () => {
+    await expect(
+      executeCodegen(getCodegenOptions('duplicate-operations.ts')),
+    ).rejects.toThrowError(/Not all operations have an unique name: layout/i);
```

Comment:
To the whole application. It's a limitation of the existing TS-Operation plugin (the codegen plugin that generates types). I think it kind of makes sense since the plugin doesn't know how to name your types otherwise to avoid conflicts. It could append some kind of filepath to them but then it wouldn't be obvious when you want to import them...